enhancements/two-nodes-openshift/2no.md

```diff
@@ -0,0 +1,501 @@
+
+Upon a peer failure, the RHEL-HA components on the surivor will fence the peer and use the OCF script to restart etcd as a new cluster of one.
+
+Upon a network failure, the RHEL-HA components ensure that exactly one node will survive, fence it's peer, and use the OCF script to restart etcd as a new cluster of one.
+
+In both cases, the control-plane will be unresponsive until etcd has been restarted.
+
+Upon rebooting, the RHEL-HA components ensure that a node remains inert (not running cri-o, kubelet, or etcd) until it sees it's peer.
+If the peer is likely to remain offline for an extended period of time, admin confirmation is required to allow the node to start OpenShift.
+
+When starting etcd, the OCF script will use etcd's cluster ID and version counter to determine whether the existing data directory can be reused, or must be erased before joining an active peer.
+
+### Workflow Description
+
+#### Cluster Creation
+
+Creation of a two node control plane will be possible via the core installer (with an additional bootstrap node), and via the Assisted Installer (without an additional bootstrap node).
+
+In the case of the core OpenShift installer, the user-facing proceedure is unchanged from a standard "IPI" installation, other than the configuration of 2 nodes instead of 3.
+Internally, the RedFish details for each node will need to make their way into the RHEL-HA configuration, but this is information already required for bare-metal hosts.
+
+In the case of the Assisted Installer, the user-facing proceedure follows the standard flow except for the configuration of 2 nodes instead of 3, and the collection of RedFish details for each node which are needed for the RHEL-HA configuration.
+
+Everything else about cluster creation will be an opaque implementation detail not exposed to the user. 
+
+#### Day 2 Proceedures
+
+As per a standard 3-node control plane, OpenShift upgrades and `MachineConfig` changes can not be applied when the cluster is in a degraded state.
+Such operations will only proceed when both peers are online and healthy.
+
+The experience of managing a 2-node control plane should be largely indistinguishable from that of a 3-node one.
+
+### API Extensions
+
+Initially the creation of an etcd cluster will be driven in the same way as other platforms.
+Once the cluster has two members, the etcd daemon will be removed from the static pod definition and recreated as a resource controlled by RHEL-HA.
+At this point, the Cluster Etcd Operator (CEO) will be made aware of this change so that some membership management functionality that is now handled by RHEL-HA can be disabled.
+This will be achieved by having the same entity that drives the configuration of RHEL-HA use the OpenShift API to update a field on the Infrastructure CR - which can only succeed if the control-plane is healthy.
+
+To enable this flow, we propose the addition of a `externallyManagedEtcd` field to the `BareMetalPlatformSpec` which defaults to False.
```

Comment:
great idea!

Reply:
By tying this to the `BareMetalPlatformSpec`, are we saying that this will only ever be supported by customers using the BareMetalPlatform? While it's more invasive, I think there is an argument to be made here about introducing an ETCD-specific config API.
Perhaps less controversial and simpler would be enabling this configuration via the FeatureGate API.

Reply:
If we look at the motivation and goals, there's no real justification for this topology in cloud environments.
So there needs to be physical machines involved, the question to me is whether `BareMetalPlatformSpec` is compatible with the Assisted flow.

Reply:
Assisted can install `platform:baremetal`.

Reply:
BTW API approvers will not let you add a boolean field :slightly_smiling_face:

Reply:
fixed